pkg/cmd/server/bootstrappolicy/controller_policy.go

```diff
@@ -163,6 +163,11 @@ func init() {
 			rbac.NewRule("create").Groups(kAuthzGroup).Resources("subjectaccessreviews").RuleOrDie(),
 			rbac.NewRule("update").Groups(templateGroup).Resources("templateinstances/status").RuleOrDie(),
 			rbac.NewRule("update").Groups(templateGroup).Resources("templateinstances/finalizers").RuleOrDie(),
+			// template instance controller needs to be able to create and get any resource the user is allowed to create.
+			rbac.NewRule("create").Groups(rbac.APIGroupAll).Resources(rbac.ResourceAll).RuleOrDie(),
```

Comment:
Wow, that's broad.

Reply:
yeah, this allows escalation flows like:

* user authorized to `create builds` can now create custom builds (passes `create builds` SAR check, template service broker passes `create builds/custom` check during actual create)
* user authorized to `create roles` or `create rolebindings` can now create or bind a role containing any creation permission

Reply:
like i said, let the debate begin :)

the alternative I can see is to switch to using impersonation instead of SAR checks, which also has the advantage of fixing some other potential issues @liggitt noted, where the user might have permission to create X, but isn't allowed to create X when field Y is Z (e.g. privileged pods).  Today that's not an issue since the controller SA can't create that either, but if we ever have a case where the controller SA can pass a particular admission control that the user can't pass, the SAR check isn't sufficient anymore and impersonation would be a better solution.

Reply:
i'm not familiar w/ cluster role aggregation?

Reply:
> i'm not familiar w/ cluster role aggregation?

It allows the injection of addition policy rules into your role by adding and removing other roles.  See https://kubernetes.io/docs/admin/authorization/rbac/#aggregated-clusterroles

Reply:
> ok...what role would we aggregate here? we still need a role that allows creation of arbitrary things.

Not arbitrary unless you really need arbitrary.  For CRDs that want to be created, they simply add a well-known label of your choosing.  You'd probably choose to aggregate `admin`.  I don't think that "*" is suitable since add-ons can declaratively indicate that you should be able to create their resources.

Reply:
That would mean granting those new permissions to everyone with "admin".  That doesn't seem like a desirable model.  (Nor, in general, does requiring the cluster-admin to go modify permissions to make Templates do what they are supposed to be able to do, strike me as very user friendly).

Also what is behind the dislike of impersonation?  It seems like a better solution than what we've done already, nevermind the growing complexity/limitations as we are now seeing as we try to extend it.

Reply:
> That would mean granting those new permissions to everyone with "admin". That doesn't seem like a desirable model.

That is the only model that makes any sense from a normal user perspective.  Consider if a broker creates a CRD and no user has access to it.  The CRD is basically useless at that point.  Without aggregation, there is also no way for the user to create a role / binding to the CRD.  Thus they basically have to use the TSB as an escalation enabler.

> (Nor, in general, does requiring the cluster-admin to go modify permissions to make Templates do what they are supposed to be able to do, strike me as very user friendly).

Seems like the exact UX I would want as a cluster admin.  Normal users do not get to "install" new APIs into the cluster and say they "own" those new APIs.  The cluster admin gets to opt-in to that.  Afterwards, users are free to create instances of those CRDs just like an other resource they have access to.

> Also what is behind the dislike of impersonation? It seems like a better solution than what we've done already, nevermind the growing complexity/limitations as we are now seeing as we try to extend it.

Because it turns the TSB into a component that is an indirect cluster admin (forever).  In general we do not have a good way to state "do X as user Y but do not let me escalate into permissions I do not already have - i.e. make sure RBAC / admission / etc would let both of us do this thing".

Reply:
> That is the only model that makes any sense from a normal user perspective. Consider if a broker creates a CRD and no user has access to it. The CRD is basically useless at that point. 

Not if i only want some subset of my users to have access to it.  

> Seems like the exact UX I would want as a cluster admin. Normal users do not get to "install" new APIs into the cluster and say they "own" those new APIs. The cluster admin gets to opt-in to that. Afterwards, users are free to create instances of those CRDs just like an other resource they have access to.

Again you're assuming I want all users in my cluster to be able to do this thing.  If I don't want that, then I need to grant the TSB a different role (not admin) so i can grant access on a more fine grained level.  This is a mess.  The TSB should not need special permissions to create things, it should be able to create anything the *user* is allowed to create.  Hence why impersonation is the right solution.

> Because it turns the TSB into a component that is an indirect cluster admin (forever). In 

So? As long as the TSB doesn't take any action that isn't impersonating the user requesting the action, who cares?

Reply:
> Not if i only want some subset of my users to have access to it.

If you want a more nuanced flow, then I am fine with you having to do it manually.  I agree with @deads2k's assessment that the TSB is a `a user focused tool` and not `a cluster-admin focused tool`.

> So? As long as the TSB doesn't take any action that isn't impersonating the user requesting the action, who cares?

Right, as long as the TSB has no bugs and there no unforeseen paths to cause it misbehave (or gain access to its token), it is perfectly fine.  I want less components with `root` on the cluster, not more.

Reply:
> If you want a more nuanced flow, then I am fine with you having to do it manually.

I understand where you're coming from, but here's how that statement comes across to me:  "i'm fine with @bparees spending his day explaining to customers how to make this work because we wouldn't let him/couldn't give him a way to do it that would have just worked out of the box."

>  I agree with @deads2k's assessment that the TSB is a a user focused tool and not a cluster-admin focused tool.

It's precisely because it's user focused that a cluster-admin should not have to go manage the TSB to enable a user to do things they can already do with their user account.  The TSB is just another api for creating resources.

It's pretty easy for me to envision a cluster administrator creating a class of users who can create a particular CRD instance(e.g. people allowed to instantiate a particular operator), and those users wanting to be able to create those CRD instances via a template.  With your proposal the administrator would have to understand they need to go explicitly grant said permission to the TSB SA (rather than just granting it to the new class of user).  And they're not going to realize that (even if we doc it) until it doesn't work.  And then they're going to forget to do it when they stand up their second cluster.  It's "just one more thing" on top of the way too many things we already have.

> Right, as long as the TSB has no bugs and there no unforeseen paths to cause it misbehave (or gain access to its token), it is perfectly fine. I want less components with root on the cluster, not more.

I know, and ideally you'd like the cluster to be on a hard drive that has been ripped out of the PC and melted down for scrap and that would make all the potential security holes go away.  But we also have to provide a good user experience.